DESIGN.md

```diff
@@ -0,0 +1,67 @@
+
+
+# Design Principles - Addressed
+
+This is a rough accounting of the design problems addressed so far.
+
+### Axis Mundi
+
+The Postgres database is the center of the world and the application design flows outward.
+
+- We want the database to maintain as much consistency as possible, such that any program interacting with the database (not just the new ODB) will see an accurate view of the world and will only be able to make modifications that are "legal" under some definition of well-formedness.
+- We want the database to be queryable easily, such that a read-replica can be made available to science staff for their own use. To this end some *checked* denormalization (such as program ids being broken into site, year, etc.) is ok.
+- Similarly astronomical units need to be in a science-friendly format, and we may need server-side functions for doing computations. We need to be cautious about pushing too much *business* logic into the database but support for domain-specific data types is probably reasonable.
+- Enumerated types like filters, dispersers, etc., are important for reporting and consistency checking and need to be defined in the database. We use code-generation to read these in at compile time and make them available as proper enumerated types in our Scala code.
+- Events should be logged to the database and echoed to traditional file logs in case database connectivity fails.
+
+The schema now represents the following notions:
+
+- Programs
+  - Title
+  - Structured ID
+- Users
+  - Name, Email, Password, Staff flag
+  - Relationship to a given program (PI, Science Contact, etc.)
+- Observations
+  - Structured ID, Name
+- Steps
+  - Bias, Dark, GCal, Science
+  - Instrument-specific (F2 only so far)
+- Enumerated Types
+  - Site
+  - Program Type
+  - Program Role (for users)
+  - Instrument
+  - GCal config (Filter, Lamp)
+  - F2 config (Disper, Filter, FPU, Lyot Wheel)
+
+The data model more or less reflects the above, and some read/write database access is provided but it's not totally built out. Basically just what's needed for import and simple queries.
+
+### Data Model and Level of Detail
+
+We need a straightforward way to deal with *portions* of the data model (for instance a list of programs without any observation information, for an "Open" dialog) without a proliferation of specialized model classes.
+
+- Model classes are parameterized on their children; i.e., any type that is represented by a foreign key in the schema. A `Program[Observation[...]]` is a program with a list of observations, a bit like the current model; a`Program[Observation.Id]` is a program with a list of observation *ids*; a `Program[Nothing]` has no observation information at all.
+- This keeps the model representation simple and flexible for dealing with the varying levels of detail needed for the UI, and means we don't have to load and ship any more information than necessary.
```

Comment:
On the seqexec side we are discussing if we transfer the whole model every time there is a change or only events that change. It sounds that `gem` is leaning to only send the minimal amount of data and send edits compatible with CRDT, is that right?